### PR TITLE
fix(appset): git files generator in matrix generator produces no params

### PR DESCRIPTION
**Pull request by crenshaw-dev**
Most of this PR is tests. The actual fix is two lines. 

The problem was introduced in 2.6.5 when we started [using field-wise templating in generators nested under `matrix`](https://github.com/argoproj/argo-cd/pull/12287). That PR fixed a significant issue where the second generator in a matrix generator wasn't being properly interpolated with templated values.

However, the fix uses a new templating function with subtly different behavior: instead of populating slice fields with nils, it populated them with empty slices.

Since we were using `appSetGenerator.Git.Directories != nil` to check whether a git generator was for directories instead of files, the bugfix above caused file generators to be incorrectly treated as directory generators.

tl;dr - a bugfix in 2.6.5 caused an issue where, if a matrix generator's second child was a files generator, the generator produced no parameters. This PR fixes that issue.

### CODE REVIEW

**Review by jannfis — APPROVED** [review]
LGTM

**Comment by crenshaw-dev** [comment]
Cherry-picked onto release-2.5 for 2.5.15.
